[Bug]: KafkaRoller fail to return the result

### Bug Description

in method rollingRestart，when order the nodes by calling podOperations.isReady, then isReady throw some exception，the code do not handle it well， it should shutdown the singleExecutor and fail the result. so the reconciliation will be nerver
 terminate.  the next reconciliation will fail to acquire lock.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Strimzi version

0.36.1

### Kubernetes version

Kubernetes 1.22

### Installation method

_No response_

### Infrastructure

_No response_

### Configuration files and logs

_No response_

### Additional context

_No response_

I add some code to catch the exception：
![image](https://github.com/strimzi/strimzi-kafka-operator/assets/6326847/7fc5f2c7-f7a2-4cf4-ab85-151a67b649ed)

the log output：
![image](https://github.com/strimzi/strimzi-kafka-operator/assets/6326847/550da490-daab-47ab-99c7-bc89a585bf52)

the excepction detail is：
[2023-08-30 13:44:25]
content: 2023-08-30 13:44:25 INFO  KafkaRoller:269 - Reconciliation #877(watch) Kafka(ali-strimzi-kafka/alikafka-post-cn-jxgffjofjfhelr1): sort nodes:[alikafka-post-cn-jxgffjofjfhelr1-32-2000-2000/2000, alikafka-post-cn-jxgffjofjfhelr1-32-2000-2001/2001] error

+1
[2023-08-30 13:44:25]
content: io.fabric8.kubernetes.client.KubernetesClientException: Operation: [get]  for kind: [Pod]  with name: [alikafka-post-cn-jxgffjofjfhelr1-32-2000-2001]  in namespace: [ali-strimzi-kafka]  failed.

+2
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:159) ~[io.fabric8.kubernetes-client-api-6.7.2.jar:?]

+3
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:187) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+4
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:141) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+5
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:92) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+6
[2023-08-30 13:44:25]
content: 	at io.strimzi.operator.common.operator.resource.AbstractReadyNamespacedResourceOperator.isReady(AbstractReadyNamespacedResourceOperator.java:65) ~[io.strimzi.operator-common-0.36.1-0.0.1-DEV.jar:0.36.1-0.0.1-DEV]

+7
[2023-08-30 13:44:25]
content: 	at io.strimzi.operator.cluster.operator.resource.KafkaRoller.lambda$rollingRestart$4(KafkaRoller.java:264) ~[io.strimzi.cluster-operator-0.36.1-0.0.1-DEV.jar:0.36.1-0.0.1-DEV]

+8
[2023-08-30 13:44:25]
content: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]

+9
[2023-08-30 13:44:25]
content: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]

+10
[2023-08-30 13:44:25]
content: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]

+11
[2023-08-30 13:44:25]
content: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]

+12
[2023-08-30 13:44:25]
content: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]

+13
[2023-08-30 13:44:25]
content: 	at java.lang.Thread.run(Thread.java:833) ~[?:?]

+14
[2023-08-30 13:44:25]
content: Caused by: java.io.IOException: /192.168.168.166:42996: GOAWAY received

+15
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:515) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+16
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:535) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+17
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleGet(OperationSupport.java:478) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+18
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleGet(BaseOperation.java:741) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+19
[2023-08-30 13:44:25]
content: 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:185) ~[io.fabric8.kubernetes-client-6.7.2.jar:?]

+20
[2023-08-30 13:44:25]
content: 	... 10 more

+21
[2023-08-30 13:44:25]
content: Caused by: java.io.IOException: /192.168.168.166:42996: GOAWAY received

+22
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection.handleGoAway(Http2Connection.java:1008) ~[java.net.http:?]

+23
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection.handleConnectionFrame(Http2Connection.java:873) ~[java.net.http:?]

+24
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection.processFrame(Http2Connection.java:748) ~[java.net.http:?]

+25
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.frame.FramesDecoder.decode(FramesDecoder.java:155) ~[java.net.http:?]

+26
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection$FramesController.processReceivedData(Http2Connection.java:232) ~[java.net.http:?]

+27
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection.asyncReceive(Http2Connection.java:674) ~[java.net.http:?]

+28
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection$Http2TubeSubscriber.processQueue(Http2Connection.java:1310) ~[java.net.http:?]

+29
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205) ~[java.net.http:?]

+30
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149) ~[java.net.http:?]

+31
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230) ~[java.net.http:?]

+32
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:303) ~[java.net.http:?]

+33
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:256) ~[java.net.http:?]

+34
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection$Http2TubeSubscriber.runOrSchedule(Http2Connection.java:1328) ~[java.net.http:?]

+35
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection$Http2TubeSubscriber.onNext(Http2Connection.java:1354) ~[java.net.http:?]

+36
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.Http2Connection$Http2TubeSubscriber.onNext(Http2Connection.java:1288) ~[java.net.http:?]

+37
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SSLTube$DelegateWrapper.onNext(SSLTube.java:210) ~[java.net.http:?]

+38
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.onNext(SSLTube.java:492) ~[java.net.http:?]

+39
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.onNext(SSLTube.java:295) ~[java.net.http:?]

+40
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run1(SubscriberWrapper.java:316) ~[java.net.http:?]

+41
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run(SubscriberWrapper.java:259) ~[java.net.http:?]

+42
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205) ~[java.net.http:?]

+43
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149) ~[java.net.http:?]

+44
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230) ~[java.net.http:?]

+45
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:303) ~[java.net.http:?]

+46
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:256) ~[java.net.http:?]

+47
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SubscriberWrapper.outgoing(SubscriberWrapper.java:232) ~[java.net.http:?]

+48
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SubscriberWrapper.outgoing(SubscriberWrapper.java:198) ~[java.net.http:?]

+49
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SSLFlowDelegate$Reader.processData(SSLFlowDelegate.java:444) ~[java.net.http:?]

+50
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SSLFlowDelegate$Reader$ReaderDownstreamPusher.run(SSLFlowDelegate.java:268) ~[java.net.http:?]

+51
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205) ~[java.net.http:?]

+52
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149) ~[java.net.http:?]

+53
[2023-08-30 13:44:25]
content: 	at jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230) ~[java.net.http:?]

--- a/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java
+++ b/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java
@@ -251,7 +251,7 @@ public class KafkaRoller {
             } catch (Exception e)   {
                 // If anything happens, we have to raise the error otherwise the reconciliation would get stuck
                 // Its logged at upper level, so we just log it at debug here
-                LOGGER.warnCr(reconciliation, "Something went wrong when trying to do a rolling restart", e);
+                LOGGER.debugCr(reconciliation, "Something went wrong when trying to do a rolling restart", e);
                 singleExecutor.shutdown();
                 result.fail(e);
             }

--- a/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java
+++ b/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java
@@ -216,37 +216,45 @@ public class KafkaRoller {
         this.podNeedsRestart = podNeedsRestart;
         Promise<Void> result = Promise.promise();
         singleExecutor.submit(() -> {
-            LOGGER.debugCr(reconciliation, "Verifying cluster pods are up-to-date.");
-            List<NodeRef> pods = new ArrayList<>(nodes.size());
+            try {
+                LOGGER.debugCr(reconciliation, "Verifying cluster pods are up-to-date.");
+                List<NodeRef> pods = new ArrayList<>(nodes.size());
 
-            for (NodeRef node : nodes)  {
-                // Order the nodes unready first otherwise repeated reconciliations might each restart a pod
-                // only for it not to become ready and thus drive the cluster to a worse state.
+                for (NodeRef node : nodes) {
+                    // Order the nodes unready first otherwise repeated reconciliations might each restart a pod
+                    // only for it not to become ready and thus drive the cluster to a worse state.
 
-                // TODO: In KRaft mode, We currently roll only nodes with the broker role. This is because of Kafka
-                //       limitations. Once managing controller nodes is supported with Kafka Admin API, this should be
-                //       fixed. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/8593.
-                if (node.broker()) {
-                    pods.add(podOperations.isReady(namespace, node.podName()) ? pods.size() : 0, node);
-                }
-            }
-            LOGGER.debugCr(reconciliation, "Initial order for updating pods (rolling restart or dynamic update) is {}", pods);
-
-            List<Future<Void>> futures = new ArrayList<>(nodes.size());
-            for (NodeRef node : pods) {
-                futures.add(schedule(node, 0, TimeUnit.MILLISECONDS));
-            }
-            Future.join(futures).onComplete(ar -> {
-                singleExecutor.shutdown();
-                try {
-                    if (allClient != null) {
-                        allClient.close(Duration.ofSeconds(30));
+                    // TODO: In KRaft mode, We currently roll only nodes with the broker role. This is because of Kafka
+                    //       limitations. Once managing controller nodes is supported with Kafka Admin API, this should be
+                    //       fixed. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/8593.
+                    if (node.broker()) {
+                        pods.add(podOperations.isReady(namespace, node.podName()) ? pods.size() : 0, node);
                     }
-                } catch (RuntimeException e) {
-                    LOGGER.debugCr(reconciliation, "Exception closing admin client", e);
                 }
-                vertx.runOnContext(ignored -> result.handle(ar.map((Void) null)));
-            });
+                LOGGER.debugCr(reconciliation, "Initial order for updating pods (rolling restart or dynamic update) is {}", pods);
+
+                List<Future<Void>> futures = new ArrayList<>(nodes.size());
+                for (NodeRef node : pods) {
+                    futures.add(schedule(node, 0, TimeUnit.MILLISECONDS));
+                }
+                Future.join(futures).onComplete(ar -> {
+                    singleExecutor.shutdown();
+                    try {
+                        if (allClient != null) {
+                            allClient.close(Duration.ofSeconds(30));
+                        }
+                    } catch (RuntimeException e) {
+                        LOGGER.debugCr(reconciliation, "Exception closing admin client", e);
+                    }
+                    vertx.runOnContext(ignored -> result.handle(ar.map((Void) null)));
+                });
+            } catch (Exception e)   {
+                // If anything happens, we have to raise the error otherwise the reconciliation would get stuck
+                // Its logged at upper level, so we just log it at debug here
+                LOGGER.warnCr(reconciliation, "Something went wrong when trying to do a rolling restart", e);
+                singleExecutor.shutdown();
+                result.fail(e);
+            }
         });
         return result.future();
     }

--- a/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaRollerTest.java
+++ b/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaRollerTest.java
@@ -7,6 +7,7 @@ package io.strimzi.operator.cluster.operator.resource;
 import io.fabric8.kubernetes.api.model.Pod;
 import io.fabric8.kubernetes.api.model.PodBuilder;
 import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.strimzi.api.kafka.model.KafkaResources;
 import io.strimzi.operator.cluster.KafkaVersionTestUtils;
 import io.strimzi.operator.cluster.model.NodeRef;
@@ -566,6 +567,24 @@ public class KafkaRollerTest {
                 asList(0),
                 KafkaRoller.FatalProblem.class, "Error while waiting for restarted pod c-kafka-0 to become ready",
                 asList(0));
+    }
+
+    @Test
+    public void testFailWhenPodIsReadyThrows(VertxTestContext testContext) throws InterruptedException {
+        PodOperator podOps = mockPodOps(podId ->
+                (podId == 0) ? failedFuture(new KubernetesClientException("Failed")) : succeededFuture()
+        );
+
+        TestingKafkaRoller kafkaRoller = new TestingKafkaRoller(null, null, addPodNames(REPLICAS),
+                podOps,
+                noException(), null, noException(), noException(), noException(),
+                brokerId -> succeededFuture(true),
+                false, null, false, null, 1);
+
+        doFailingRollingRestart(testContext, kafkaRoller,
+                List.of(),
+                KubernetesClientException.class, "Failed",
+                List.of());
     }
 
     private TestingKafkaRoller rollerWithControllers(PodOperator podOps, int... controllers) {

